service/storage_proxy.cc

```diff
@@ -2222,11 +2243,36 @@ storage_proxy::get_paxos_participants(const sstring& ks_name, const dht::token &
  */
 future<> storage_proxy::mutate(std::vector<mutation> mutations, db::consistency_level cl, clock_type::time_point timeout, tracing::trace_state_ptr tr_state, service_permit permit, bool raw_counters) {
     if (_cdc && _cdc->needs_cdc_augmentation(mutations)) {
-        return _cdc->augment_mutation_call(timeout, std::move(mutations), tr_state).then([this, cl, timeout, tr_state, permit = std::move(permit), raw_counters](std::tuple<std::vector<mutation>, lw_shared_ptr<cdc::operation_result_tracker>>&& t) mutable {
-            auto mutations = std::move(std::get<0>(t));
-            auto tracker = std::move(std::get<1>(t));
-            return _mutate_stage(this, std::move(mutations), cl, timeout, std::move(tr_state), std::move(permit), raw_counters, std::move(tracker));
+        // Separate counters from non-counters for CDC
+        auto cnt_begin = boost::range::partition(mutations, [] (const mutation& m) {
+            return !m.schema()->is_counter();
         });
+        // Counter "shards" mutations are not augmented. Counter "delta" mutations are augmented upon applying on leader.
+        if (cnt_begin != mutations.end()) {
+            // There are some counter mutations, so isolate them from augmenting
+            std::vector<mutation> counter_mutations;
+            counter_mutations.insert(counter_mutations.end(),
+                    std::make_move_iterator(cnt_begin),
+                    std::make_move_iterator(mutations.end()));
+            mutations.erase(cnt_begin, mutations.end());
+
+            return _cdc->augment_mutation_call(timeout, std::move(mutations), tr_state).then(
+                    [this, cl, timeout, tr_state, permit = std::move(permit), raw_counters, cmut = std::move(counter_mutations)]
+                    (std::tuple<std::vector<mutation>, lw_shared_ptr<cdc::operation_result_tracker>>&& t) mutable {
+                auto mutations_with_cdc = std::move(std::get<0>(t));
+                auto tracker = std::move(std::get<1>(t));
+                cmut.insert(cmut.end(),
+                        std::make_move_iterator(mutations_with_cdc.begin()),
+                        std::make_move_iterator(mutations_with_cdc.end()));
```

Comment:
this is a lot of cdc-specific logic cluttering this `mutate` function whose purpose should be simple to explain: "augment the set of mutations if needed by adding more mutations (e.g. cdc-related), then send those mutations".

The whole "augment" logic should be moved out to a separate function which doesn't know about the "send" part.

Reply:
Well, one could argue that `mutate` does not hit close to "augment the set of mutations if needed by adding more mutations (e.g. cdc-related), then send those mutations". But otherwise I agree.